File: solutions/ruby/difference-of-squares/1/difference_of_squares.rb
class Squares
  def initialize(n)
    @n = n
  end

  def square_of_sum
    sum = (@n * (@n+1)) / 2
    sum * sum
  end

  def sum_of_squares
    (@n * (@n + 1) * (2 * @n + 1)) / 6
  end

  def difference
    square_of_sum - sum_of_squares
  end
end

File: solutions/clojure/difference-of-squares/1/src/difference_of_squares.clj
(ns difference-of-squares)

(defn square-of-sum
  "Returns the square of the sum of the numbers up to N."
  [N]
  (let [S (/ (* N (+ N 1)) 2)]
    (* S S)))

(defn sum-of-squares
  "Returns the sum of the squares of the numbers up to N."
  [N]
  (/ (* N (+ N 1) (+ (* 2 N) 1)) 6)
  )

(defn difference
  "Returns the difference between the square of the sum
  and the sum of the squares of the numbers up to N."
  [N]
  (- (square-of-sum N) (sum-of-squares N)))
